### PR TITLE
feat: report HNSW duplicate statistics

### PR DESCRIPTION
**Pull request by Spockkk0225**
## Summary

- estimate duplicate vectors with a lightweight hash set during HNSW bulk build
- always report duplicate count and percentage when creating the index
- add an in-memory regression test and document duplicate-vector performance guidance

## Testing

- `cmake --build build --target neug_vector_search_extension -j4`
- `NEUG_RUN_EXTENSION_TESTS=1 PYTHONPATH=build/tools/python_bind:tools/python_bind /opt/homebrew/bin/python3.14 -m pytest tools/python_bind/tests/test_hnsw_index.py::test_bulk_build_reports_duplicate_vectors -q`

Related to #956.

### CODE REVIEW

File: extension/vector_search/src/hnsw_index.cc [hunk 1]
```diff
@@ -336,21 +339,39 @@ Status HNSWIndex::BulkBuild(const VertexSet& vertices) {
     return Status::RuntimeError(
         "HNSWIndex must be open and bound before bulk build");
   }
+  flat_hash_set<uint64_t> duplicate_statistics;
+  const auto vector_byte_size = dimension_ * sizeof(float);
```
Comment (shirly121) [inline extension/vector_search/src/hnsw_index.cc]: 这里最好不要假设 float 类型，可以加一些判断，未来可能扩展其他类型

File: extension/vector_search/src/hnsw_index.cc [hunk 2]
```diff
@@ -352,6 +361,18 @@ Status HNSWIndex::BulkBuild(const VertexSet& vertices) {
...
+      index_id_accessor_->GetDeletedIndexIDs().size();
+  const auto duplicate_vector_count =
+      indexed_vector_count - duplicate_statistics.size();
+  LOG(WARNING) << "HNSW duplicate statistics for index '" << meta_->name
+               << "': " << duplicate_vector_count << " / "
+               << indexed_vector_count << " ("
+               << (indexed_vector_count == 0
+                       ? 0.0
+                       : 100.0 * duplicate_vector_count / indexed_vector_count)
+               << "%) duplicate vectors";
```
Comment (longbinlai) [inline extension/vector_search/src/hnsw_index.cc]: **这条 warning 在默认 Python 配置下会被整条丢掉，功能对用户不可见。**（blocking）

`tools/python_bind/src/neug_binding.cc` 的 `setup_logging()` 在 `PYBIND11_MODULE` 里就被调用（第 154 行），没有 `DEBUG=1` 时设的是：

```cpp
FLAGS_minloglevel = 2;      // 2 for error, 3 for fatal
FLAGS_logtostderr = false;  // Log to file instead of stderr
```

`minloglevel=2` 会把 WARNING（level 1）直接丢弃，连日志文件都不写。extension 是 `PRIVATE neug` 链接的（`extension/vector_search/CMakeLists.txt:29`），glog 是同一份实例，所以这个 flag 对它同样生效。

用户会怎么踩坑：普通 Python 用户执行 `CREATE INDEX` 之后什么提示都看不到，必须先 `DEBUG=1` 再重跑一遍建索引才能看到统计。而 issue #956 明确要求 "The warning should be available through both Cypher and the Python API" —— 现在 Cypher 结果和 Python API 两条路都拿不到，issue 的诉求没被满足。

建议把统计通过 `CREATE INDEX` 的返回结果/notice 暴露出来；如果这轮先不做，至少要在文档里写明必须 `DEBUG=1` 才看得到，否则这个功能等于没上线。

File: extension/vector_search/src/hnsw_index.cc [hunk 3]
```diff
@@ -352,6 +361,18 @@ Status HNSWIndex::BulkBuild(const VertexSet& vertices) {
           ". See logs for details.");
     }
   }
+  const auto indexed_vector_count =
+      index_id_accessor_->GetNextIndexID() -
+      index_id_accessor_->GetDeletedIndexIDs().size();
+  const auto duplicate_vector_count =
+      indexed_vector_count - duplicate_statistics.size();
```
Comment (longbinlai) [inline extension/vector_search/src/hnsw_index.cc]: **分母取了全局 index id 空间，会报出并不存在的重复。**（blocking）

分子 `duplicate_statistics.size()` 是本次循环里实际哈希到的 distinct 数，分母却是全局计数，这两个不是同一个集合：

`GetNextIndexID()` 返回的是所有 clone 共享的原子计数器 —— `include/neug/storages/index/index_id_accessor.h` 里的注释写得很清楚：*"Clones share this counter so index IDs allocated by aborted transactions are not reused by later transactions."* 本快照真正可见的上界是 `visible_limit_`，而其他事务/已回滚事务消耗掉的 id **只会在本 accessor 下一次 `UpsertVID` 时才被补录进 `deleted_index_ids_`**（`src/storages/index/index_id_accessor.cc` 的 `UpsertVID`，"Interleaved accessors may have consumed IDs since this snapshot's visible boundary"）。

所以 `GetNextIndexID() - deleted.size()` 会比本快照真实存活的向量数偏大，重复数被凭空放大 —— 极端情况下一条重复都没有也会报出重复。

同一个文件的查询路径就是这么区分两者的：`HNSWIndex` 搜索时先取 `GetVisibleLimit()`，再用 `visible_limit < GetNextIndexID()` 判断"有 id 已分配但本快照不可见"。统计这里漏掉了这一层。

建议直接在循环里自己数（`++hashed_count`），重复数 = `hashed_count - duplicate_statistics.size()`。语义变成"我实际哈希过的这些向量里有多少重复"，既准确，也顺带避免了 unsigned 相减下溢打出 `18446744073709551615 / 100 (1.8e+19%)` 这种数字。

File: extension/vector_search/src/hnsw_index.cc [hunk 4]
```diff
@@ -352,6 +361,18 @@ Status HNSWIndex::BulkBuild(const VertexSet& vertices) {
...
+  LOG(WARNING) << "HNSW duplicate statistics for index '" << meta_->name
```
Comment (longbinlai) [inline extension/vector_search/src/hnsw_index.cc]: **0 重复时也会刷一条 WARNING。**（建议）

这里是无条件 `LOG(WARNING)`，意味着每次 `CREATE INDEX` 都产出一条 WARNING，哪怕重复率是 0%。用户按 WARNING 级别 grep 日志排查问题时，健康的索引也会被这条命中，看起来像出了故障。

建议低于阈值（比如 20%）时用 `LOG(INFO)`，超过阈值才升到 `LOG(WARNING)` —— 真正需要用户注意的才是 warning。

File: extension/vector_search/src/hnsw_index.cc [hunk 5]
```diff
@@ -336,14 +339,20 @@ Status HNSWIndex::BulkBuild(const VertexSet& vertices) {
     return Status::RuntimeError(
         "HNSWIndex must be open and bound before bulk build");
   }
+  flat_hash_set<uint64_t> duplicate_statistics;
+  const auto vector_byte_size = dimension_ * sizeof(float);
```
Comment (longbinlai) [inline extension/vector_search/src/hnsw_index.cc]: **CPU 开销可以接受，但内存开销没有上限也没有开关。**（建议）

先说 CPU：`std::hash<std::string_view>` 底层是 murmur2/cityhash（libstdc++ 是 MurmurHash64A，libc++ 是 `__murmur2_or_cityhash`），吞吐在 3–6 GB/s 量级。按此估算 dim=768（3072 B）约 0.6–1 µs/向量，dim=128 约 0.1–0.2 µs，dim=8 只有几个 ns。对比同一循环里的 `AddWithSource`（`m=16, ef_construction=200`，每次插入几百到上千次距离计算，dim=768 每次约 30–70 ns，合计 15–35 µs），哈希占 2–5%；而且哈希是 O(N)、建图是 O(N·log N)，数据越大占比越小。哈希读的又正好是紧接着 `AddWithSource` 要用的那块内存，带宽上近似免费。所以这块不用担心。

真正的成本是内存：`flat_hash_set<uint64_t>` 每条约 9 B slot，abseil 容量是 2 的幂、最大负载 7/8，实际 9–18 B/distinct 向量。1 亿条不重复向量 → 额外 1–1.8 GB 峰值，而且没有任何开关能关掉。低维时比例更难看：dim=8 的向量本体只有 32 B，哈希集反而占到向量数据的一半以上。

issue #956 自己就留了口子："If detecting every duplicate would make index construction too expensive, a sampled warning or index-build statistic would still be useful." 建议要么采样、要么加个可关闭的选项，别让所有人无条件付这笔内存。

File: tools/python_bind/tests/test_hnsw_index.py [hunk 6]
```diff
@@ -308,6 +308,30 @@ def test_update_and_delete_maintain_index(advanced_connection):
...
+                parameters={"id": i, "embedding": vector},
+            )
+
+        capfd.readouterr()
+        conn.execute(
+            "CREATE INDEX item_hnsw ON Item USING HNSW (embedding) "
+            "WITH (metric='ip', m=16, ef_construction=200);"
+        )
+        warning = capfd.readouterr().err
+        assert "83 / 100 (83%) duplicate vectors" in warning
```
Comment (longbinlai) [inline tools/python_bind/tests/test_hnsw_index.py]: **这个断言依赖 `DEBUG=1`，而 CI 没有设，这条测试在 CI 里应该会挂。**（blocking）

按 `neug_binding.cc` 的 `setup_logging()`，没有 `DEBUG=1` 时 `FLAGS_minloglevel = 2`，`LOG(WARNING)` 压根不会产生，stderr 上自然什么都没有。而 `.github/workflows/neug-extension-test.yml` 跑的是：

```
NEUG_RUN_EXTENSION_TESTS=true python3 -m pytest -sv tests/test_hnsw_index.py
```

没有 `DEBUG`。PR body 里贴的本地命令也没有，估计是本地 shell 里恰好带着。

（顺带排除一个可能的误判：`-s` 不影响 `capfd`，我在 pytest 8.3.5 下单独验过，`capfd` 在 `--capture=no` 时仍能抓到 fd 2 的输出，所以问题不在这里。）

建议要么在这条测试里显式设 `DEBUG=1` 重开 Database（注意 `setup_logging()` 是 import 时执行的，进程内改 env 不生效，可能得起子进程），要么等统计改成走查询返回值之后直接断言返回值 —— 后者顺便把上面那条"用户看不到"的问题一起解决了。

File: tools/python_bind/tests/test_hnsw_index.py [hunk 7]
```diff
@@ -308,6 +308,30 @@ def test_update_and_delete_maintain_index(advanced_connection):
     assert rows[0][0] == 334
 
 
+def test_bulk_build_reports_duplicate_vectors(capfd):
```
Comment (longbinlai) [inline tools/python_bind/tests/test_hnsw_index.py:318]: **测试只覆盖了"有重复"这一种情况，两个会出错的场景都没测到。**（建议）

`i % 17` 造 17 个 distinct 值断言 `83 / 100 (83%)`，算得对，而且刚好是整数百分比、不会被 ostream 默认 6 位有效精度弄成 `83.3333` —— 这点选得挺讲究。

缺的两条：

1. **0 重复的用例**：补上就会暴露"每次 CREATE INDEX 都刷 WARNING"的噪音行为，也能锁住修好之后的日志级别。
2. **先有回滚事务或并发写分配过 index id，再 `CREATE INDEX`**：这正是分母那条问题会算错的场景，现在没有任何测试能拦住它。

File: doc/source/extensions/vector_search.md [hunk 8]
```diff
@@ -319,6 +319,21 @@ Therefore, users can choose either workflow:
 - Workflow 1: Import Data First, Build Index Later
 - Workflow 2: Create Index First, Continuously Write Data
 
+### Duplicate Vectors
+
+A large number of duplicate vectors can degrade HNSW index construction and
+query performance. When `CREATE INDEX` bulk-builds an HNSW index, NeuG estimates
+the duplicate rate from vector hashes and writes a warning containing approximate
+duplicate statistics, for example:
+
```
Comment (longbinlai) [inline doc/source/extensions/vector_search.md:328]: **文档承诺的行为和默认实际行为不符。**（blocking，跟代码那条是同一件事）

这里写 NeuG "writes a warning containing approximate duplicate statistics"，但默认的 Python 配置下 `FLAGS_minloglevel = 2` 会把这条 WARNING 整条丢掉（见 `neug_binding.cc` 的 `setup_logging()`），用户既不会在 stderr 看到、也不会在日志文件里看到。照文档做完 `CREATE INDEX` 什么都没出现，很容易被当成 bug 报上来。

要么把统计改成通过查询结果暴露（issue #956 要的就是这个），要么在这段里明确写出需要 `DEBUG=1` 才能看到，并给出示例命令。

另外建议补一句：这个统计只在 `CREATE INDEX` 批量建索引时计算，后续 `CREATE`/`SET` 增量写入的向量不参与统计 —— 不然用户会以为它是持续生效的。

File: extension/vector_search/src/hnsw_index.cc (hunk 1 above)
Comment (longbinlai) [reply extension/vector_search/src/hnsw_index.cc]: 补两条证据，供参考（不重复你的结论，只是把现成的东西指出来）：

1. **今天这里是安全的**：`Rebind` 已经硬性拒绝了非 float 数组 —— `if (element_type != DataTypeId::kFloat) return Status(ERR_INVALID_ARGUMENT, "HNSWIndex supports only FLOAT arrays")`，而 `BulkBuild` 必然在 `Rebind` 之后跑（`storage_index.h` 里写明契约是 `Init -> Open -> Rebind -> BulkBuild`）。所以现在不会算错，属于未来扩展类型时的隐患。

2. **有现成的值可以直接用，不用自己判断类型**：`HNSWVecSource` 构造时已经按 array 的 child type 算好了 `vector_byte_size_`（`case DataTypeId::kFloat: vector_byte_size_ = sizeof(float) * dimension;`，非 float 直接 `THROW_NOT_SUPPORTED_EXCEPTION`），正是这里要的字节数。把它暴露成一个 getter、这里直接取，比 `dimension_ * sizeof(float)` 更省事，将来加类型也自动跟上，不需要在两处各写一遍 switch。

File: extension/vector_search/src/hnsw_index.cc (hunk 3 above)
Comment (shirly121) [reply extension/vector_search/src/hnsw_index.cc]: 这里建议直接按照点的数量来统计重复率，不要按照 index_count，确实会有被废弃的无意义的 index_id

File: extension/vector_search/src/hnsw_index.cc (hunk 2 above)
Comment (Spockkk0225) [reply extension/vector_search/src/hnsw_index.cc]: 默认 Python 下 warning 实际可见，原因是 glog 以静态库链接，Python binding 和 vector-search 扩展各自包含独立的 FLAGS_minloglevel。setup_logging() 修改的是 binding 自己的 flags，无法影响扩展中的 LOG(WARNING)。扩展加载时再次出现 Logging before InitGoogleLogging() 也印证了这一点。

File: extension/vector_search/src/hnsw_index.cc (hunk 5 above)
Comment (Spockkk0225) [reply extension/vector_search/src/hnsw_index.cc]: 时间开销：相比于向量插入的巨大开销，执行哈希计算的时间可以忽略不计，且缓存命中率较高，整体仅通过string_view包装，没有额外的拷贝开销，因此这块保持不变。
内存开销：每个向量只需要在哈希表中存储一个8B的哈希值，目前应用场景都是高维向量（>128维），相比于构建向量本身的瞬时峰值开销，这部分内存占用比例也比较小，而且只会在构建阶段一次性使用，构建完毕后就会释放。

File: tools/python_bind/tests/test_hnsw_index.py (hunk 6 above)
Comment (Spockkk0225) [reply tools/python_bind/tests/test_hnsw_index.py]: 同上，默认 Python 下 warning 实际可见。

File: extension/vector_search/src/hnsw_index.cc (hunk 1 above)
Comment (Spockkk0225) [reply extension/vector_search/src/hnsw_index.cc]: 已改用vector_byte_size_
https://github.com/alibaba/neug/pull/975/commits/1db555bb1f9608b3dc549202bd9fe3c7640df81e

File: extension/vector_search/src/hnsw_index.cc (hunk 3 above)
Comment (Spockkk0225) [reply extension/vector_search/src/hnsw_index.cc]: 改为在循环中直接统计count
https://github.com/alibaba/neug/pull/975/commits/9fb7abf2ed81f6c65f7a792f66f3664cdbaf9355

File: extension/vector_search/src/hnsw_index.cc (hunk 4 above)
Comment (Spockkk0225) [reply extension/vector_search/src/hnsw_index.cc]: 低于20%的重复率改为INFO级别的LOG
https://github.com/alibaba/neug/pull/975/commits/d97339455ae1ecd9c4d932ec6e3d2d15c7bee77d

File: doc/source/extensions/vector_search.md (hunk 8 above)
Comment (Spockkk0225) [reply doc/source/extensions/vector_search.md:328]: 已经补充文档，这个统计量只针对初始创建索引，后续更新不会计算。
https://github.com/alibaba/neug/pull/975/commits/f51f97d7effd553e5adc76885f4d79fa8ac28964

File: tools/python_bind/tests/test_hnsw_index.py (hunk 7 above)
Comment (Spockkk0225) [reply tools/python_bind/tests/test_hnsw_index.py:318]: 已补充两个测试：重复率较低时日志级别为INFO，有回滚时重复率正确计算
https://github.com/alibaba/neug/pull/975/commits/91c4625024046eddd3a54418f1f45520eeade245
